### ISSUE TITLE
Property-based tests (proptest) for algorithm mathematical invariants

### ISSUE THREAD

**Issue by dollspace-gay** [dollspace-gay/ferrolearn#11]
## Summary

Add proptest harnesses that verify ML algorithm outputs satisfy their mathematical invariants for randomly generated inputs — bridging the gap between Kani's bounded proofs and sklearn equivalence tests.

## Context

Kani proves properties for small bounded inputs. Sklearn equivalence tests prove agreement on fixed datasets. Neither covers "do mathematical invariants hold on arbitrary realistic-sized inputs?" — that's what property-based testing addresses.

### Regressors
| Property | Applies to |
|---|---|
| `predict(fit(X, y), X)` approximates `y` (training R2 > 0 for non-degenerate data) | All regressors |
| `coef_.len() == n_features` | LinearRegression, Ridge, Lasso, ElasticNet |
| Ridge with alpha=0 matches OLS | Ridge vs LinearRegression |
| ElasticNet with l1_ratio=0 matches Ridge | ElasticNet vs Ridge |
| ElasticNet with l1_ratio=1 matches Lasso | ElasticNet vs Lasso |
| Lasso produces sparse coefficients (some coef ~= 0 for large alpha) | Lasso |

### Classifiers
| Property | Applies to |
|---|---|
| `predict()` output labels are subset of training labels | All classifiers |
| `predict_proba()` rows sum to 1.0 | LogisticRegression, DecisionTreeClassifier, GaussianNB |
| `predict_proba()` values in [0, 1] | LogisticRegression, DecisionTreeClassifier, GaussianNB |
| `predict()` agrees with `argmax(predict_proba())` | All with predict_proba |
| `classes_` matches unique training labels | All classifiers |
| DecisionTree with max_depth=1 produces <= 3 leaf nodes | DecisionTreeClassifier |

### Transformers
| Property | Applies to |
|---|---|
| `inverse_transform(transform(X))` approximately recovers X | PCA (within component subspace), StandardScaler (exact) |
| StandardScaler output has mean ~= 0, std ~= 1 | StandardScaler |
| PCA components are orthonormal | PCA |
| `explained_variance_ratio_` sums to <= 1.0 | PCA |
| PCA `transform()` output has n_components columns | PCA |

### Clustering
| Property | Applies to |
|---|---|
| `labels_` values in [0, n_clusters) | KMeans |
| `cluster_centers_.shape == (n_clusters, n_features)` | KMeans |
| `predict(cluster_centers_) == [0, 1, ..., k-1]` (centers predict their own cluster) | KMeans |
| `inertia_ >= 0` | KMeans |

### Metrics (complement to Kani)
| Property | Applies to |
|---|---|
| `accuracy_score(y, y) == 1.0` (perfect prediction) | accuracy_score |
| `r2_score(y, y) == 1.0` | r2_score |
| `mse(y, y) == 0.0` | mean_squared_error |
| Metric symmetry where applicable | adjusted_rand_score |
| Triangle inequality properties | distance-based metrics |

## Implementation

```rust
use proptest::prelude::*;

proptest! {
    #[test]
    fn predict_proba_sums_to_one(
        x in array_strategy(10..50, 2..10),
        y in label_strategy(10..50, 2..5),
    ) {
        let model = LogisticRegression::<f64>::new();
        let fitted = model.fit(&x, &y).unwrap();
        let proba = fitted.predict_proba(&x).unwrap();
        for row in proba.rows() {
            let sum: f64 = row.sum();
            prop_assert!((sum - 1.0).abs() < 1e-10);
        }
    }
}
```

Shared test helpers for generating valid random datasets (non-degenerate, sufficient rank, reasonable scale).

## Acceptance criteria
- [ ] proptest harnesses for every property listed above
- [ ] Custom strategies for generating valid ML datasets (non-degenerate, mixed scales)
- [ ] All tests pass with default proptest config (256 cases)
- [ ] Regressions from any found failures are added as fixed test cases
